src/security/Handshake.cc

```diff
@@ -274,39 +274,47 @@ Security::HandshakeParser::parseModernRecord()
     Must(record.fragment.length() || record.type == ContentType::ctApplicationData);
 
     if (currentContentType != record.type) {
+        tkMessages.expectMore(false);
         parseMessages();
         Must(tkMessages.atEnd()); // no currentContentType leftovers
-        fragments = record.fragment;
+        tkMessages.reset(record.fragment, true);
         currentContentType = record.type;
     } else {
-        fragments.append(record.fragment);
+        tkMessages.reset(tkMessages.leftovers().append(record.fragment), true);
     }
 
-    if (tkRecords.atEnd() && !done)
+    if (!done)
         parseMessages();
 }
 
 /// parses one or more "higher-level protocol" frames of currentContentType
 void
 Security::HandshakeParser::parseMessages()
 {
-    tkMessages.reset(fragments, false);
-    for (; !tkMessages.atEnd(); tkMessages.commit()) {
-        switch (currentContentType) {
-        case ContentType::ctChangeCipherSpec:
-            parseChangeCipherCpecMessage();
-            continue;
-        case ContentType::ctAlert:
-            parseAlertMessage();
-            continue;
-        case ContentType::ctHandshake:
-            parseHandshakeMessage();
-            continue;
-        case ContentType::ctApplicationData:
-            parseApplicationDataMessage();
-            continue;
+    try {
+        for (; !tkMessages.atEnd(); tkMessages.expectMore(false), tkMessages.commit()) {
+            switch (currentContentType) {
+                case ContentType::ctChangeCipherSpec:
+                    parseChangeCipherCpecMessage();
+                    continue;
+                case ContentType::ctAlert:
+                    parseAlertMessage();
+                    continue;
+                case ContentType::ctHandshake:
+                    parseHandshakeMessage();
+                    continue;
+                case ContentType::ctApplicationData:
+                    parseApplicationDataMessage();
+                    continue;
+            }
+            skipMessage("unknown ContentType msg [fragment]");
         }
-        skipMessage("unknown ContentType msg [fragment]");
+    }
+    catch (const Parser::BinaryTokenizer::InsufficientInput &) {
+        tkMessages.rollback();
```

Comment:
Please move this rollback() call to the start of the messages parsing (i.e. the beginning of this method). Otherwise, we are creating an impression that it is important to catch all InsufficuentInput exceptions to roll back the tokenizer -- something that is very easy to forget/miss because a lot of code may throw... In reality, we only need to roll back at the _beginning_ of a parsing sequence, which usually has a much more well-defined/specific code location. This sketch illustrates a typical sequence for parsing multiple things:

```C++
tkMessages.rollback();
while (!done)
    ... parse one message ...
```

Please note that the above sketch also implies that the termination condition of this loop should not be atEnd()! There are two reasons to change that condition:

1. One of the benefits of having an InsufficientInput exception is that it is safe to parse anything without checking atEnd() -- the parser will do the right thing when there are not enough bytes. While it is not wrong to stop parsing atEnd(), doing so creates a false impression that we should check atEnd() before parsing _anything_ or "things may break". That false assumption creates terrible code like this:
 
```C++
if (!tk.atEnd())
    a = t.parseA();

if (!tk.atEnd()
    b = tk.parseB();
else
    ...
 ```

2. If we stop atEnd(), then parseMessages() (or, in the current PR code, the loop) has four outcomes instead of three:

* parsing error (exception)
* insufficient input (exception)
* success (no exception; got everything we wanted)
* insufficient input (no exception)

Removing the last bullet simplifies code logic a lot!

Reply:
I followed your sketch and removed entirely the try/catch.